prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,163 @@ def check_for_location_conflicts(substations):
         )
 
 
+def map_lines_to_substations_using_coords(substations, lines, rounding=3):
+    """Map lines to substations using coordinates
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param int rounding: coordinates rounding. Default to 3.
+    :return: (*pandas.DataFrame*) -- indices: lines id, columns: lines endpoint
+        coordinates (*FROM* and *TO*) and id of substations found at these coordinates (*FROM_SUB* and *TO_SUB*)
+    """
+    subcoord2subid = (
+        substations.round(rounding).groupby(["LONGITUDE", "LATITUDE"]).groups
+    )
+    print(
+        f"initial number of substations: {substations.shape[0]}\n"
+        f"substations with same location after rounding: "
+        f"{substations.shape[0] - len(subcoord2subid)}\n"
+    )
+
+    lines_coord_from = lines["COORDINATES"].map(
+        lambda x: (round(x[0][0], rounding), round(x[0][1], rounding))
+    )
+    lines_coord_to = lines["COORDINATES"].map(
+        lambda x: (round(x[-1][0], rounding), round(x[-1][1], rounding))
+    )
+    line2sub = lines_coord_from.to_frame(name="FROM").join(
+        lines_coord_to.to_frame(name="TO"), how="inner"
```

Comment:
Do we need to specify `how="inner"`? I think the indices of the two should be identical, so the `how` parameter shouldn't matter.

Reply:
Done